### PR TITLE
[zh-cn]: Update translation of String.replace()

### PR DESCRIPTION
**Pull request by JasonLamv-t**
### Description

Same as the title.

### Motivation

Ditto.

### Additional details

None.

### Related issues and pull requests

#13975

### CODE REVIEW

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 1]
```diff
@@ -214,16 +243,12 @@ console.log(retArr);
...
-- 从 Gecko 39 开始，`flags` 参数便被弃用，且在调用时会抛出一个控制台警告（[Firefox bug 1142351](https://bugzil.la/1142351)）。
-- 从 Gecko 47 开始，在非发行版本中已不再支持非标准的 `flags` 参数，并且不久后会完全移除该参数（[Firefox bug 1245801](https://bugzil.la/1245801)）。
-- 从 Gecko 49 开始，不再支持非标准的 `flags` 参数（[Firefox bug 1108382](https://bugzil.la/1108382)）。
-
 ## 参见
 
-- {{jsxref("String.prototype.match()")}}
-- {{jsxref("RegExp.prototype.exec()")}}
-- {{jsxref("RegExp.prototype.test()")}}
+- [在 `core-js` 中，`String.prototype.replace` 的 Polyfill 实现修复了一些问题，并支持现代行为，比如 `Symbol.replace` 的支持。](https://github.com/zloirock/core-js#ecmascript-string-and-regexp)
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:248]:
```suggestion
- [在 `core-js` 中，`String.prototype.replace` 的 Polyfill 实现修复了一些问题，并支持现代行为，比如 `Symbol.replace` 的支持](https://github.com/zloirock/core-js#ecmascript-string-and-regexp)
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 2]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+| -------- | ------------------------------------------------------------------- |
+| `$$`     | 插入一个 `$`。                                                      |
+| `$&`     | 插入匹配的子字符串。                                                |
+| `` $` `` | 插入匹配子字符串之前的字符串片段。                                  |
+| `$'`     | 插入匹配子字符串之后的字符串片段。                                  |
+| `$n`     | 插入第 `n`（从 1 开始计数）个捕获组，其中 `n` 是小于 100 的正整数。 |
+| `$`      | 插入名称为 `Name` 的命名组。                                        |
 
-你可以指定一个函数作为第二个参数。在这种情况下，当匹配执行后，该函数就会执行。函数的返回值作为替换字符串。 (注意：上面提到的特殊替换参数在这里不能被使用。) 另外要注意的是，如果第一个参数是正则表达式，并且其为全局匹配模式，那么这个方法将被多次调用，每次匹配都会被调用。
+只有当 `pattern` 参数是一个 {{jsxref("RegExp")}} 对象时，`$n` 和 `$<Name>` 才可用。如果 `pattern` 是字符串，或者相应的捕获组在正则表达式中不存在，则该模式将被替换为一个字面字符串。如果该组存在但未匹配（因为它是一个分支的一部分），则将用空字符串替换它。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:60]:
```suggestion
只有当 `pattern` 参数是一个 {{jsxref("RegExp")}} 对象时，`$n` 和 `$<Name>` 才可用。如果 `pattern` 是字符串，或者相应的捕获组在正则表达式中不存在，则该模式将被替换为一个字面量。如果该组存在但未匹配（因为它是一个分支的一部分），则将用空字符串替换它。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 3]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+
+```js
+"foo".replace(/(f)/, "$2");
+// "$2oo"; 正则表达式没有第二个组
+
+"foo".replace("f", "$1");
+// "$1oo"; pattern 是一个字符串，所以它没有任何组
+
+"foo".replace(/(f)|(g)/, "$2");
+// "oo"; 第二个组存在但未匹配
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:70]:
```suggestion
// "$2oo"；正则表达式没有第二个组

"foo".replace("f", "$1");
// "$1oo"；pattern 是一个字符串，所以它没有任何组

"foo".replace(/(f)|(g)/, "$2");
// "oo"；第二个组存在但未匹配
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 4]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
-| `match`           | 匹配的子串。（对应于上述的$&。）                                                                                                                                                                                   |
-| `p1,p2, ...`      | 假如 replace() 方法的第一个参数是一个{{jsxref("RegExp")}} 对象，则代表第 n 个括号匹配的字符串。（对应于上述的$1，$2 等。）例如，如果是用 `/(\a+)(\b+)/` 这个来匹配，`p1` 就是匹配的 `\a+`，`p2` 就是匹配的 `\b+`。 |
-| `offset`          | 匹配到的子字符串在原字符串中的偏移量。（比如，如果原字符串是 `'abcd'`，匹配到的子字符串是 `'bc'`，那么这个参数将会是 1）                                                                                           |
-| `string`          | 被匹配的原字符串。                                                                                                                                                                                                 |
-| NamedCaptureGroup | 命名捕获组匹配的对象                                                                                                                                                                                               |
+你可以将函数指定为第二个参数。在这种情况下，匹配完成后将调用该函数。函数的结果（返回值）将用作替换字符串。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:75]:
```suggestion
你可以将第二个参数指定为函数。在这种情况下，匹配完成后将调用该函数。函数的结果（返回值）将用作替换字符串。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 5]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+  return replacement;
+}
+```
+
+该函数的参数如下所示：
+
+- `match`
+  - : 匹配的子字符串。（对应于上面的 `$&`。）
+- `p1, p2, …, pN`
+  - : 如果 `replace()` 的第一个参数是 {{jsxref("RegExp")}} 对象，则为捕获组（包括命名捕获组）找到的第 `n` 个字符串。（对应于上面的 `$1`，`$2` 等。）例如，如果 `pattern` 是 `/(\d+)(\w+)/`，则 `p1` 是 `\a+` 的匹配项，`p2` 是 `\b+` 的匹配项。如果该组是析取的一部分（例如 `"abc".replace(/(a)|(b)/, Replacer)`），则不匹配的替代项将为 `undefined`。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:92]: 上下文统一

```suggestion
  - : 如果 `replace()` 的第一个参数是 {{jsxref("RegExp")}} 对象，则为捕获组（包括命名捕获组）找到的第 `n` 个字符串。（对应于上面的 `$1`、`$2` 等。）例如，如果 `pattern` 是 `/(\d+)(\w+)/`，则 `p1` 是 `\a+` 的匹配项，`p2` 是 `\b+` 的匹配项。如果该组是分支的一部分（例如 `"abc".replace(/(a)|(b)/, Replacer)`），则不匹配的替代项将为 `undefined`。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 6]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+  - : 一个捕获组命名组成的对象，值是匹配的部分（如果没有匹配，则为 `undefined`）。仅在 `pattern` 包含至少一个命名捕获组时才存在。
+
+参数的确切数量取决于第一个参数是否为 {{jsxref("RegExp")}} 对象，以及它有多少个捕获组。
+
+以下示例将设置 `newString` 为 `'abc - 12345 - #$*%'`：
 
 ```js
 function replacer(match, p1, p2, p3, offset, string) {
-  // p1 is nondigits, p2 digits, and p3 non-alphanumerics
+  // p1 是非数字，p2 是数字，且 p3 非字母数字字符
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:106]: 加上“字符”感觉怪怪的，“p3 不是字符？”

```suggestion
  // p1 是非数字，p2 是数字，且 p3 非字母数字
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 7]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
-var newstr = str.replace(/xmas/i, "Christmas");
+const str = "Twas the night before Xmas...";
+const newstr = str.replace(/xmas/i, "Christmas");
 console.log(newstr); // Twas the night before Christmas...
 ```
 
-### 在 `replace()` 中使用 `global` 和 `ignore` 选项
+这将打印 `'Twas the night before Christmas...'`。
+
+> **注意：** 有关正则表达式的更多解释，请参阅[正则表达式指南](/zh-CN/docs/Web/JavaScript/Guide/Regular_Expressions)。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:129]:
```suggestion
> **备注：** 有关正则表达式的更多解释，请参阅[正则表达式指南](/zh-CN/docs/Web/JavaScript/Guide/Regular_Expressions)。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 8]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+
+### 在 replace() 中使用 global 和 ignoreCase 标志
 
 下面的例子中，正则表达式包含有全局替换 (g) 和忽略大小写 (i) 的选项，这使得 replace 方法用'oranges'替换掉了所有出现的"apples".
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:134]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 9]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+const newstr = str.replace(re, "oranges");
+console.log(newstr); // oranges are round, and oranges are juicy.
 ```
 
+这将打印 `'oranges are round, and oranges are juicy'`。
+
 ### 交换字符串中的两个单词
 
-下面的例子演示了如何交换一个字符串中两个单词的位置，这个脚本使用$1 和 $2 代替替换文本。
+以下脚本交换字符串中的单词。对于替换文本，脚本使用[捕获组](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions/Groups_and_backreferences)和 `$1` 和 `$2` 替换模式。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:148]:
```suggestion
以下脚本交换字符串中的单词。对于替换文本，脚本使用[捕获组](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions/Groups_and_backreferences)以及 `$1` 和 `$2` 替换模式。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 10]
```diff
@@ -5,207 +5,236 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+假设我们想创建一个替换器，将偏移数据附加到每个匹配的字符串中。因为替换函数已经接收到 `offset` 参数，如果正则表达式是静态已知的，那么这将变得很简单。
 
-_x_x___x___x___
+```js
+"abcd".replace(/(bc)/, (match, p1, offset) => `${match} (${offset}) `);
+// "abc (1) d"
 ```
 
-**输出：**
+然而，如果我们希望这个替换器能够适用于任何正则表达式模式，那么它将很难泛化。替换器是*可变*参数的，它接收的参数数量取决于存在的捕获组数量。我们可以使用 [rest 参数](/zh-CN/docs/Web/JavaScript/Reference/Functions/rest_parameters)，但它也会将 `offset`、`string` 等收集到数组中。根据正则表达式的特性，`groups` 可能会被传递或者不会被传递，这也使得很难泛化地知道哪个参数对应于 `offset`。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:211]:
```suggestion
然而，如果我们希望这个替换器能够适用于任何正则表达式模式，那么它将很难泛化。替换器是*可变*参数的，它接收的参数数量取决于存在的捕获组数量。我们可以使用[剩余参数参数](/zh-CN/docs/Web/JavaScript/Reference/Functions/rest_parameters)，但它也会将 `offset`、`string` 等收集到数组中。根据正则表达式的特性，`groups` 可能会被传递或者不会被传递，这也使得很难泛化地知道哪个参数对应于 `offset`。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 11]
```diff
@@ -214,16 +243,12 @@ console.log(retArr);
...
-- 从 Gecko 39 开始，`flags` 参数便被弃用，且在调用时会抛出一个控制台警告（[Firefox bug 1142351](https://bugzil.la/1142351)）。
-- 从 Gecko 47 开始，在非发行版本中已不再支持非标准的 `flags` 参数，并且不久后会完全移除该参数（[Firefox bug 1245801](https://bugzil.la/1245801)）。
-- 从 Gecko 49 开始，不再支持非标准的 `flags` 参数（[Firefox bug 1108382](https://bugzil.la/1108382)）。
-
 ## 参见
 
-- {{jsxref("String.prototype.match()")}}
-- {{jsxref("RegExp.prototype.exec()")}}
-- {{jsxref("RegExp.prototype.test()")}}
+- [在 `core-js` 中，`String.prototype.replace` 的 Polyfill 实现修复了一些问题，并支持现代行为，比如 `Symbol.replace` 的支持](https://github.com/zloirock/core-js#ecmascript-string-and-regexp)
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:248]:
```suggestion
- [`core-js` 中 `String.prototype.replace` 的 Polyfill，修复了一些问题，并实现了现代行为，比如 `Symbol.replace` 的支持](https://github.com/zloirock/core-js#ecmascript-string-and-regexp)
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 12]
```diff
@@ -5,207 +5,234 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
-- `regexp` (pattern)
-  - : 一个{{jsxref("RegExp")}} 对象或者其字面量。该正则所匹配的内容会被第二个参数的返回值替换掉。
-- `substr` (pattern)
-  - : 一个将被 `newSubStr` 替换的 {{jsxref("String","字符串")}}。其被视为一整个字符串，而不是一个正则表达式。仅第一个匹配项会被替换。
-- `newSubStr` (replacement)
-  - : 用于替换掉第一个参数在原字符串中的匹配部分的{{jsxref("String", "字符串")}}。该字符串中可以内插一些特殊的变量名。参考下面的[使用字符串作为参数](#使用字符串作为参数)。
-- `function` (replacement)
-  - : 一个用来创建新子字符串的函数，该函数的返回值将替换掉第一个参数匹配到的结果。参考下面的[指定一个函数作为参数](#指定一个函数作为参数)。
+- `pattern`
+  - : 可以是字符串或者一个带有 [`Symbol.replace`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/replace) 方法的对象，典型的例子就是[正则表达式](/zh-CN/docs/Web/JavaScript/Guide/Regular_Expressions)。任何没有 `Symbol.replace` 方法的值都会被强制转换为字符串。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:21]:
```suggestion
  - : 可以是字符串或者一个带有 [`Symbol.replace`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/replace) 方法的对象，典型的例子就是[正则表达式](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions)。任何没有 `Symbol.replace` 方法的值都会被强制转换为字符串。
```

File: files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md [hunk 13]
```diff
@@ -5,207 +5,234 @@ slug: Web/JavaScript/Reference/Global_Objects/String/replace
...
+const str = "Twas the night before Xmas...";
+const newstr = str.replace(/xmas/i, "Christmas");
 console.log(newstr); // Twas the night before Christmas...
 ```
 
-### 在 `replace()` 中使用 `global` 和 `ignore` 选项
+这将打印 `'Twas the night before Christmas...'`。
 
-下面的例子中，正则表达式包含有全局替换 (g) 和忽略大小写 (i) 的选项，这使得 replace 方法用'oranges'替换掉了所有出现的"apples".
+> **备注：** 有关正则表达式的更多解释，请参阅[正则表达式指南](/zh-CN/docs/Web/JavaScript/Guide/Regular_Expressions)。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/global_objects/string/replace/index.md:129]:
```suggestion
> **备注：** 有关正则表达式的更多解释，请参阅[正则表达式指南](/zh-CN/docs/Web/JavaScript/Guide/Regular_expressions)。
```

**Review by yin1999 — APPROVED** [review]
Thanks

**Pull request merged**